### PR TITLE
Fix loading of old ConcatTransform models

### LINKED ISSUE #1300: Bug in ConcatTransform legacy loading
When loading a ConcatTransform with an older version, we need to read an additional int, for sizeof(float) that used to be serialized with the transform.

### PR DESCRIPTION
**Pull request by yaeldekel**
Fixes #1300 .

### CODE REVIEW

File: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs [hunk 1]
```diff
@@ -306,6 +306,10 @@ private ColumnInfo[] LoadLegacy(ModelLoadContext ctx)
             //          int: index of the alias
             //          int: string id of the alias
             //      int: -1, marks the end of the list
+            // int: sizeof(Float).
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: >int: sizeof(Float). [](start = 15, length = 19)

why is it at the end?

File: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs [hunk 2]
```diff
@@ -306,6 +306,10 @@ private ColumnInfo[] LoadLegacy(ModelLoadContext ctx)
...
+
+            var sizeofFloat = ctx.Reader.ReadInt32();
+            Contracts.CheckDecode(sizeofFloat == 4 || sizeofFloat == 8);
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: >4 || sizeofFloat == 8 [](start = 49, length = 21)

`sizeof(float)` and `sizeof(double)` ?

Comment (Zruty0) [reply src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: Actually, how about we disallow the double build? It's not like anyone was using it or something

---
In reply to: [226423915](https://github.com/dotnet/machinelearning/pull/1301#discussion_r226423915) [](ancestors = 226423915)

**Comment by Zruty0** [comment]
>         private ColumnInfo[] LoadLegacy(ModelLoadContext ctx)

I wonder if we should add a test for the legacy load. Maybe not one transform at a time, but a bigger model with a bunch of transforms?

---
Refers to: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs:295 in 2bd0a6f. [](commit_id = 2bd0a6fca670476df574478eb0ce601396b894e1, deletion_comment = False)

**Comment by eerhardt** [comment]
>         private ColumnInfo[] LoadLegacy(ModelLoadContext ctx)

No matter what - we should add at least 1 test here in this PR. You can follow the same approach I took for #1290 and add the model to [https://github.com/dotnet/machinelearning-testdata/tree/master/Microsoft.ML.TestModels/TestModels](https://github.com/dotnet/machinelearning-testdata/tree/master/Microsoft.ML.TestModels/TestModels)

---
In reply to: [431120918](https://github.com/dotnet/machinelearning/pull/1301#issuecomment-431120918) [](ancestors = 431120918)

---
Refers to: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs:295 in 2bd0a6f. [](commit_id = 2bd0a6fca670476df574478eb0ce601396b894e1, deletion_comment = False)

**Pull request merged**